### PR TITLE
Initial support for FreeBSD

### LINKED ISSUE #2917: Support for FreeBSD
### Describe the feature you'd like supported

I could successfully build msquic cloned yesterday on FreeBSD 13.1-RELEASE.
It may affect other platforms and I haven't checked yet if it really works, but it compiled and output libmsquic.so.2 anyway.

By the way, as I'm not very familiar with cmake, I manually changed build.make, etc so FreeBSD uses 
datapath_kqueue.c not datapath_epoll.c.
FreeBSD is basically similar to Mac OS X and kqueue should be used for event polling.

### Proposed solution

Here is the patch (not very organized though):
[msquic.freebsd.patch.txt](https://github.com/microsoft/msquic/files/9221576/msquic.freebsd.patch.txt)

### Additional context

It will be very helpful for many people if msquic officially supports FreeBSD.

Thank you.

### PR DESCRIPTION
**Pull request by scopedog**
## Description

This at least builds and outputs libmsquic.so on FreeBSD 13.1-RELEASE (the latest release). Closes #2917.

## Testing

I haven't tested if it really works yet. Tests needed. I will start testing soon.

### CODE REVIEW

**Comment by nibanks** [comment]
Looks like this broke all macOS builds:
```
/Users/runner/work/1/s/src/platform/datapath_kqueue.c:37:2: error: "can't determine socket option"
#error "can't determine socket option"
 ^
/Users/runner/work/1/s/src/platform/datapath_kqueue.c:1243:36: error: use of undeclared identifier 'DSTADDR_SOCKOPT'
            if (CMsg->cmsg_type == DSTADDR_SOCKOPT) {
                                   ^
/Users/runner/work/1/s/src/platform/datapath_kqueue.c:2172:31: error: use of undeclared identifier 'DSTADDR_SOCKOPT'
            CMsg->cmsg_type = DSTADDR_SOCKOPT;
                              ^
3 errors generated.
```
Could you please have a look?

**Comment by scopedog** [comment]
Isn't IP_PKTINFO defined in MacOS ?
In src/platform/datapath_kqueue.c, does the following 

```
#if defined IP_RECVDSTADDR -> #if defined(IP_RECVDSTADDR)
#elif defined IP_PKTINFO -> #elif defined(IP_PKTINFO)
```

solve the problem?

**Comment by nibanks** [comment]
I don't know. Feel free to push a change and we can let the CI test it out.

File: src/platform/datapath_kqueue.c [hunk 1]
```diff
@@ -865,7 +888,11 @@ CxPlatSocketContextInitialize(
         setsockopt(
             SocketContext->SocketFd,
             ForceIpv4 ? IPPROTO_IP : IPPROTO_IPV6,
+#if defined(IP_RECVPKTINFO)
             ForceIpv4 ? IP_RECVPKTINFO : IPV6_RECVPKTINFO,
+#elif defined(DSTADDR_SOCKOPT)
+            ForceIpv4 ? DSTADDR_SOCKOPT : IPV6_RECVPKTINFO,
+#endif
```
Comment (nibanks) [inline src/platform/datapath_kqueue.c:895]: You have to have an #else here or it will fail, right?

File: src/platform/datapath_kqueue.c [hunk 2]
```diff
@@ -1211,7 +1238,10 @@ CxPlatSocketContextRecvComplete(
                 FoundTOS = TRUE;
             }
         } else if (CMsg->cmsg_level == IPPROTO_IP) {
-            if (CMsg->cmsg_type == IP_PKTINFO) {
+#if defined(DSTADDR_SOCKOPT)
+            if (CMsg->cmsg_type == DSTADDR_SOCKOPT) {
+#else
+#endif
```
Comment (nibanks) [inline src/platform/datapath_kqueue.c:1244]: You have to have something here or the `else if` below won't work.

Comment (scopedog) [reply src/platform/datapath_kqueue.c:1244]: Well I noticed
```
// For all *NIX
#if defined(IP_RECVDSTADDR)
#define DSTADDR_SOCKOPT IP_RECVDSTADDR
#define DSTADDR_DATASIZE (CMSG_SPACE(sizeof(struct in_addr)))
#define dstaddr(x) (CMSG_DATA(x))

#elif defined(IP_PKTINFO)
#define DSTADDR_SOCKOPT IP_PKTINFO
#define DSTADDR_DATASIZE (CMSG_SPACE(sizeof(struct in_pktinfo)))
#define dstaddr(x) (&(((struct in_pktinfo *)(CMSG_DATA(x)))->ipi_addr))

#endif

```

should be moved under #include <*.h>.

Comment (scopedog) [reply src/platform/datapath_kqueue.c:1244]: Could you try the latest update?

**Comment by scopedog** [comment]
What test failed?
It is very interesting because my modification is mainly for kqueue which Linux does not use and others are supposed to affect Linux very little. Well, there may be a mistake.

**Comment by nibanks** [comment]
Looks like lots of macOS tests result in crash because of this:
```
==================================================================================
 msquicplatformte.4880.core
==================================================================================
(lldb) target create "/Users/runner/work/1/s/artifacts/bin/macos/x64_Debug_openssl/msquicplatformtest" --core "/Users/runner/work/1/s/artifacts/logs/msquicplatformtest/08.01.2022.16.40.21/DataPathTest_DataPathTest.UdpData_4/msquicplatformte.4880.core"
Core file '/Users/runner/work/1/s/artifacts/logs/msquicplatformtest/08.01.2022.16.40.21/DataPathTest_DataPathTest.UdpData_4/msquicplatformte.4880.core' (x86_64) was loaded.
(lldb) bt all
warning: could not execute support code to read Objective-C class data in the process. This may reduce the quality of type information available.
* thread #1
  * frame #0: 0x00007ff8101b53ea libsystem_kernel.dylib`__psynch_cvwait + 10
    frame #1: 0x00007ff8101efa6f libsystem_pthread.dylib`_pthread_cond_wait + 1249
    frame #2: 0x000000010cf21cf4 msquicplatformtest`CxPlatInternalEventWaitWithTimeout(Event=0x00007ff7b3021400, TimeoutMs=2000) at quic_platform_posix.h:863:18 [opt]
    frame #3: 0x000000010cef0ad0 msquicplatformtest`DataPathTest_UdpData_Test::TestBody(this=0x0000600001758900) at DataPathTest.cpp:735:5 [opt]
    frame #4: 0x000000010cf396f6 msquicplatformtest`void testing::internal::HandleExceptionsInMethodIfSupported<testing::Test, void>(object=<unavailable>, method=00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00, location="the test body")(), char const*) at gtest.cc:2659:12 [opt]
    frame #5: 0x000000010cf3966f msquicplatformtest`testing::Test::Run(this=0x0000600001758900) at gtest.cc:2676:5 [opt]
    frame #6: 0x000000010cf39fa8 msquicplatformtest`testing::TestInfo::Run(this=0x00007fc253709860) at gtest.cc:2855:11 [opt]
    frame #7: 0x000000010cf3aa95 msquicplatformtest`testing::TestSuite::Run(this=0x00007fc253709970) at gtest.cc:3014:30 [opt]
    frame #8: 0x000000010cf439ee msquicplatformtest`testing::internal::UnitTestImpl::RunAllTests(this=0x00007fc253704790) at gtest.cc:5872:44 [opt]
    frame #9: 0x000000010cf434f6 msquicplatformtest`bool testing::internal::HandleExceptionsInMethodIfSupported<testing::internal::UnitTestImpl, bool>(object=<unavailable>, method=00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00, location="auxiliary test code (environments or event listeners)")(), char const*) at gtest.cc:2659:12 [opt]
    frame #10: 0x000000010cf43450 msquicplatformtest`testing::UnitTest::Run(this=0x000000010d184e18) at gtest.cc:5446:10 [opt]
    frame #11: 0x000000010cee0191 msquicplatformtest`RUN_ALL_TESTS() at gtest.h:2293:73 [opt]
    frame #12: 0x000000010cee0099 msquicplatformtest`main(argc=<unavailable>, argv=0x00007ff7b3021850) at main.cpp:38:12 [opt]
    frame #13: 0x000000011bdd452e dyld`start + 462
  thread #2
    frame #0: 0x00007ff8101b900e libsystem_kernel.dylib`__pthread_kill + 10
    frame #1: 0x00007ff8101ef1ff libsystem_pthread.dylib`pthread_kill + 263
    frame #2: 0x00007ff81013ad24 libsystem_c.dylib`abort + 123
    frame #3: 0x000000010cf21e78 msquicplatformtest`quic_bugcheck(File="/Users/runner/work/1/s/src/platform/datapath_kqueue.c", Line=1258, Expr="FoundLocalAddr") at platform_posix.c:83:5 [opt]
    frame #4: 0x000000010cf24702 msquicplatformtest`CxPlatSocketContextRecvComplete(SocketContext=0x00007fc25370b690, BytesTransferred=1024) at datapath_kqueue.c:1258:5 [opt]
    frame #5: 0x000000010cf2503e msquicplatformtest`CxPlatSocketContextProcessEvents(Event=0x00007000078d7d40) at datapath_kqueue.c:1442:13 [opt]
    frame #6: 0x000000010cf261f9 msquicplatformtest`CxPlatDataPathRunEC(Context=0x0000600000e58410, CurThreadId=<unavailable>, WaitTime=<unavailable>) at datapath_kqueue.c:2364:13 [opt]
    frame #7: 0x000000010cf20514 msquicplatformtest`CxPlatWorkerThread(Context=0x0000600000e58370) at platform_worker.c:317:17 [opt]
    frame #8: 0x00007ff8101ef4e1 libsystem_pthread.dylib`_pthread_start + 125
    frame #9: 0x00007ff8101eaf6b libsystem_pthread.dylib`thread_start + 15
WARNING: [08/01/2022 16:41:54] DataPathTest/DataPathTest.UdpData/4 failed:
Note: Google Test filter = DataPathTest/DataPathTest.UdpData/4
```

**Comment by scopedog** [comment]
Thanks for testing.
Well, MacOS X has many features of FreeBSD and mixes with others, which is somehow confusing.
Please test with a new one.
If it doesn't pass the kqueue stuff, I'll restore the original one and create a patch that affects only FreeBSD.
By the way, current patch doesn't work on FreeBSD yet.
setsockopt(IP_PKTINFO) is a little annoying on FreeBSD.

File: src/platform/CMakeLists.txt [hunk 3]
```diff
@@ -20,11 +20,12 @@ if("${CX_PLATFORM}" STREQUAL "windows")
     endif()
 else()
     set(SOURCES ${SOURCES} inline.c platform_posix.c storage_posix.c cgroup.c)
-    if(CX_PLATFORM STREQUAL "linux")
+    if(CX_PLATFORM STREQUAL "linux" AND NOT CMAKE_SYSTEM_NAME STREQUAL "FreeBSD")
         set(SOURCES ${SOURCES} datapath_epoll.c)
     else()
         set(SOURCES ${SOURCES} datapath_kqueue.c)
     endif()
+# message(STATUS "Sources: ${SOURCES}")
```
Comment (nibanks) [inline src/platform/CMakeLists.txt:28]: please remove

File: src/platform/datapath_kqueue.c [hunk 4]
```diff
@@ -865,7 +888,11 @@ CxPlatSocketContextInitialize(
         setsockopt(
             SocketContext->SocketFd,
             ForceIpv4 ? IPPROTO_IP : IPPROTO_IPV6,
+#if defined(IP_RECVPKTINFO)
             ForceIpv4 ? IP_RECVPKTINFO : IPV6_RECVPKTINFO,
+#elif defined(DSTADDR_SOCKOPT)
```
Comment (nibanks) [inline src/platform/datapath_kqueue.c:893]:
```suggestion
#else
```
It must be defined, or you throw an error above.

File: src/platform/datapath_kqueue.c [hunk 5]
```diff
@@ -26,6 +35,20 @@
...
+#define dstaddr(x)		(&(((struct in_pktinfo *)(CMSG_DATA(x)))->ipi_addr))
+#elif defined(IP_RECVDSTADDR)
+#define DSTADDR_SOCKOPT		IP_RECVDSTADDR
+#define DSTADDR_DATASIZE	(CMSG_SPACE(sizeof(struct in_addr)))
+#define dstaddr(x)		(CMSG_DATA(x))
+#else
+#error "No socket option specified"
+#endif
+
+
```
Comment (nibanks) [inline src/platform/datapath_kqueue.c:51]:
```suggestion

```
Don't need two new lines.

File: src/platform/datapath_kqueue.c [hunk 6]
```diff
@@ -26,6 +35,20 @@
 #include "datapath_kqueue.c.clog.h"
 #endif
 
+// For all *NIX
+#if defined(IP_PKTINFO)
+#define DSTADDR_SOCKOPT		IP_PKTINFO
+#define DSTADDR_DATASIZE	(CMSG_SPACE(sizeof(struct in_pktinfo)))
+#define dstaddr(x)		(&(((struct in_pktinfo *)(CMSG_DATA(x)))->ipi_addr))
```
Comment (nibanks) [inline src/platform/datapath_kqueue.c:42]: `DSTADDR_DATASIZE` and `dstaddr` are unused, right? I so, please remove.

**Comment by scopedog** [comment]
Are the errors related to my changes?
I'd like to push a minor change in checking an interface index in datapath_kqueue.c.
Before that, I'd like to include the modification to the errors but I don't know what causes them.

**Review by nibanks — APPROVED** [review]

**Comment by scopedog** [comment]
Well, I'm currently testing some programs on FreeBSD and need to add some more modification.
How can I do that?

**Comment by nibanks** [comment]
Since I already merged the PR, you'll have to create a new PR. If you don't want a PR merged, the best thing to do is mark it as a **Draft**.

**Comment by scopedog** [comment]
OK. I'll create a new PR later. I'll close the current one anyway. Thanks

**Comment by wfurt** [comment]
BTW you may checkout https://github.com/dotnet/runtime/issues/14537 
FreeBSD port was on my TODO list as well. As .NET 7.0 should be reasonably stable, it may be worth of running Quic & HTTP/3 tests.
Also there is hook to build .pkg for macOS. `fpm` also supports FreeBSD so we may fix it as well.

**Comment by scopedog** [comment]
OK. I will check it later.

**Pull request merged**